crates/pallet-domains/src/block_tree.rs

```diff
@@ -0,0 +1,201 @@
+        // The genesis receipt is generated and added to the block tree by the runtime upon domain
+        // instantiation, thus it is unchallengeable and must always be the same.
+        ReceiptType::NewBranch if domain_block_height.is_zero() => {
+            Err(Error::ChallengeGenesisReceipt)
+        }
+        ReceiptType::NewHead
+        | ReceiptType::NewBranch
+        | ReceiptType::CurrentHead
+        | ReceiptType::Stale => Ok(()),
+    }
+}
+
+/// Process the execution receipt to add it to the block tree
+///
+/// NOTE: only `NewHead`, `NewBranch` and `CurrentHead` type of receipt is expected
+/// for this function, passing other type of receipt will result in an `UnexpectedReceiptType`
+/// error.
+pub(crate) fn process_execution_receipt<T: Config>(
+    domain_id: DomainId,
+    submitter: OperatorId,
+    execution_receipt: ExecutionReceiptOf<T>,
+    receipt_type: ReceiptType,
+) -> Result<(), Error> {
+    let er_hash = execution_receipt.hash();
+    match receipt_type {
+        er_type @ ReceiptType::NewHead | er_type @ ReceiptType::NewBranch => {
+            // Construct and add a new domain block to the block tree
+            let domain_block_height = execution_receipt.domain_block_height;
+            let domain_block = DomainBlock {
+                execution_receipt,
+                operator_ids: sp_std::vec![submitter],
+            };
+            BlockTree::<T>::mutate(domain_id, domain_block_height, |er_hashes| {
+                er_hashes
+                    .try_insert(er_hash)
+                    .map_err(|_| Error::ExceedMaxBlockTreeFork)?;
+                Ok(())
+            })?;
+            DomainBlocks::<T>::insert(er_hash, domain_block);
+
+            if er_type == ReceiptType::NewHead {
+                // Update the head receipt number
+                HeadReceiptNumber::<T>::insert(domain_id, domain_block_height);
+
+                // Prune expired domain block
+                if let Some(to_prune) =
+                    domain_block_height.checked_sub(&T::BlockTreePruningDepth::get())
+                {
+                    for block in BlockTree::<T>::take(domain_id, to_prune) {
+                        DomainBlocks::<T>::remove(block);
+                    }
+                    // Remove the block's `ExecutionInbox` as the block is pruned and not need
+                    // to verify the its receipt's `extrinsics_root` anymore
+                    ExecutionInbox::<T>::remove(domain_id, to_prune);
+                }
+            }
+        }
+        ReceiptType::CurrentHead => {
+            // Add confirmation to the current head receipt
+            DomainBlocks::<T>::mutate(er_hash, |maybe_domain_block| {
+                let domain_block = maybe_domain_block.as_mut().expect(
+                    "The domain block of `CurrentHead` receipt is checked to be exist in `execution_receipt_type`; qed"
+                );
+                domain_block.operator_ids.push(submitter);
```

Comment:
Right, there are some comments on the `operator_ids` field that mentioned this, I think it is expected because an operator can submit multiple bundles within one consensus block and all of the bundle contains the same receipt that confirms the current head receipt. And since `operator_ids` is used to reward/slash operator, the reward/slash should account for every receipt the operator submitted.

Reply:
Resolving, but nice to add a bit more context to explain why duplicated operator_ids are expected.